### PR TITLE
[lld][WebAssembly] Fix bitcode LTO order in archive parsing

### LINKED ISSUE #62243: wasm-ld crash when LTO enabled: `__cxx_global_var_init.1` present in "init functions" but not defined
We've seen certain situations where in LTO builds wasm-ld will crash during `Writer::createCallCtorsFunction` because one of the init function in the generted LTO object file (` __cxx_global_var_init.1`) in undefined and does not get assigned a function index (so crashes in `sym->getFunctionIndex()`).

I'm still working on reducing the repro case as more.

This seems to have started with https://reviews.llvm.org/D135427, which I guess makes sense since that change explicitly calls out `__cxx_global_var_init` as being discarded.

### PR DESCRIPTION
**Pull request by aheejin**
tl;dr:
When doing LTO on multiple archives, the order with which bitcodes are linked to the LTO module is hard to control, given that processing undefined symbols can lead to parsing of an object file, which in turn lead to parsing of another object file before finishing parsing of the previous file. This can result in encountering a non-prevailing comdat first when linking, which can make the the symbol undefined, and the real definition is added later with an additional prefix to avoid duplication (e.g. `__cxx_global_var_init` and `__cxx_global_var_init.2`)

So this one-line fix ensures we compile bitcodes in the order that we process comdats, so that when multiple archived bitcode files have the same variable with the same comdat, we make sure that the prevailing comdat will be linked first in the LTO.

Fixes #62243.

---

- Long version, feel free to skip:

When linking (non-archived) bitcode files directly in LTO, we add files in `bitcodeFiles` vector in the order they are given in the command line:
https://github.com/llvm/llvm-project/blob/fb57f4e0e0b302ec1b3181e952a4bd4b3c57a286/lld/wasm/Driver.cpp#L1201-L1204 -> https://github.com/llvm/llvm-project/blob/fb57f4e0e0b302ec1b3181e952a4bd4b3c57a286/lld/wasm/SymbolTable.cpp#L54

The order they are added in `bitcodeFiles` is the order they are compiled and linked into the LTO module later.

While parsing these bitcode files, we also add symbols to the symbol table. Depending on their status, they are added as defined symbols or undefined symbols.

---

On the other hand, when linking archives that contain bitcode files, the order bitcode files are added to `bitcodeFiles` vector is hard to predict.

When parsing archives, firstly, symbols are parsed as lazy symbols: https://github.com/llvm/llvm-project/blob/fb57f4e0e0b302ec1b3181e952a4bd4b3c57a286/lld/wasm/InputFiles.cpp#L734-L739

And then we handle exported symbols here:
https://github.com/llvm/llvm-project/blob/fb57f4e0e0b302ec1b3181e952a4bd4b3c57a286/lld/wasm/Driver.cpp#L1212-L1215 where we try to fetch lazy symbols' original objects: https://github.com/llvm/llvm-project/blob/fb57f4e0e0b302ec1b3181e952a4bd4b3c57a286/lld/wasm/Driver.cpp#L708 -> https://github.com/llvm/llvm-project/blob/907ed77ad1d7a154317e5f8398d17d441711dc38/lld/wasm/Symbols.cpp#L428 which causes the corresponding object file to be added to the symbol table:
https://github.com/llvm/llvm-project/blob/fb57f4e0e0b302ec1b3181e952a4bd4b3c57a286/lld/wasm/InputFiles.cpp#L763 and the object file is parsed:
https://github.com/llvm/llvm-project/blob/fb57f4e0e0b302ec1b3181e952a4bd4b3c57a286/lld/wasm/SymbolTable.cpp#L53

But this `parse` call can lead to the parsing of other files before it is added to `bitcodeFiles` in the next line. For example, we have two bitcode files `a.o` and `b.o`, each of which is archived as `liba.a` and `libb.a`. Both files have a variable with the same name and the same comdat. When handling undefined symbols, suppose we start from a symbol from `a.o`. We arrive here and all comdats in `a.o` are added to `keptComdats`:
https://github.com/llvm/llvm-project/blob/fb57f4e0e0b302ec1b3181e952a4bd4b3c57a286/lld/wasm/InputFiles.cpp#L844-L845 And we call `createBitCodeSymbol` on all symbols in `a.o`: https://github.com/llvm/llvm-project/blob/fb57f4e0e0b302ec1b3181e952a4bd4b3c57a286/lld/wasm/InputFiles.cpp#L847-L848 But this can cause the loading of another object in case some of those symbols are undefined functions or data, like here: https://github.com/llvm/llvm-project/blob/fb57f4e0e0b302ec1b3181e952a4bd4b3c57a286/lld/wasm/InputFiles.cpp#L791-L792 ->
https://github.com/llvm/llvm-project/blob/fb57f4e0e0b302ec1b3181e952a4bd4b3c57a286/lld/wasm/SymbolTable.cpp#L534

So effectively, in the middle of parsing `a.o`, we proceed to parse `b.o`. And when parsing `b.o`, we encounter the comdat variable. But because we already added that comdat in `keptComdats` when parsing `a.o`, so `b.o`'s comdat variable ends up being excluded, so it is added as an undefined data, whereas `a.o`'s comdat variable will be added as a defined data:
https://github.com/llvm/llvm-project/blob/fb57f4e0e0b302ec1b3181e952a4bd4b3c57a286/lld/wasm/InputFiles.cpp#L786-L798

For a lazy symbol, `addUndefinedData` does not replace the symbol unless it is the first time being inserted, but `addDefinedData` does: https://github.com/llvm/llvm-project/blob/fb57f4e0e0b302ec1b3181e952a4bd4b3c57a286/lld/wasm/SymbolTable.cpp#L382-L385 And because of this `a.o`'s comdat is to be the prevailing one, because the symbol's `getFile()` will return `a.o`:
https://github.com/llvm/llvm-project/blob/907ed77ad1d7a154317e5f8398d17d441711dc38/lld/wasm/LTO.cpp#L104 But because `b.o`'s parsing started in the middle of `a.o`s parsing, it ends first and gets added to `bitcodeFiles` first. https://github.com/llvm/llvm-project/blob/fb57f4e0e0b302ec1b3181e952a4bd4b3c57a286/lld/wasm/SymbolTable.cpp#L54

So to sum up, `bitcodeFiles` now contains [`b.o`, `a.o`] but `a.o` has the prevailing comdat. This does not happen when directly linking bitcodes (without archives) because symbols are added in one-pass in the order specified in the command line. This also does not happen in non-LTO (even with archives), because object parsing uses `ObjFile::parse` which is a separate process from this bitcode processing.

---

I'm not sure if there is a rule that the prevailing comdat has to exist in the first file having that comdat within `bitcodeFiles`, but my observation says this is likely the case.

When adding bitcodes to the LTO, if a symbol's comdat is not a prevailing one, this code demotes its linkage to `available_externally`: https://github.com/llvm/llvm-project/blob/23c84fb362849865990c0e160158b19f54742147/llvm/lib/LTO/LTO.cpp#L869-L895 and subsequently all other variables that are associated with the same comdat in that object file are demoted in the same way. https://github.com/llvm/llvm-project/blob/23c84fb362849865990c0e160158b19f54742147/llvm/lib/LTO/LTO.cpp#L788-L792 In out case, because of the orders of `bitcodeFiles`, we start from `b.o`, and the comdat symbols there become `available_externally`. This is how `__cxx_global_var_init` in `global_var_init2.ll` in the attached test becomes `available_externally`.

When linking those bitcodes in the LTO, if we traverse symbols in the order of prevailing->non-prevailing, we end up keeping only one global variable, but if we traverse in the other way around (i.e., non-prevailing first), we end up adding two symbols, because it lets us add a `available_externally` symbol when we don't already have a definition for it:
https://github.com/llvm/llvm-project/blob/23c84fb362849865990c0e160158b19f54742147/llvm/lib/LTO/LTO.cpp#L974-L986 Also when it is an `available_externally` symbol, it becomes a declaration, not a definition. So we end up with a module like this:
```ll
declare dso_local void @__cxx_global_var_init()
define internal void @__cxx_global_var_init.2() comdat($unused) {
  ...
}
```

---

So this one-line fix ensures we compile bitcodes in the order that we process comdats, so that when multiple archived bitcode files have the same variable with the same comdat, we make sure that the prevailing comdat will be linked first in the LTO.

---

This crash is said to be happening after https://github.com/llvm/llvm-project/commit/12050a3fb7344694cfd7527d4cca0033729bcfc5 but even reverting this is not really a solution for us because we end up with two definitions if we do that:
```ll
define internal void @__cxx_global_var_init() comdat($unused) {
  ...
}
define internal void @__cxx_global_var_init.2() comdat($unused) {
  ...
}
```
And the wasm's `__wasm_call_ctors` will be like
```wast
 (func $__wasm_call_ctors
  (call $emscripten_stack_init)
  (call $__cxx_global_var_init)
  (call $__cxx_global_var_init.2)
 )
```

### CODE REVIEW

**Comment by sbc100** [comment]
Hmm.. I wonder why I didn't get an amount when I was assigned as reviewer here.

**Comment by aheejin** [comment]
Do you get an email notification now for this comment?

**Comment by aheejin** [comment]
@sbc100

**Comment by aheejin** [comment]
test

File: lld/wasm/SymbolTable.cpp [hunk 1]
```diff
@@ -50,8 +50,10 @@ void SymbolTable::addFile(InputFile *file, StringRef symName) {
 
   // LLVM bitcode file
   if (auto *f = dyn_cast<BitcodeFile>(file)) {
-    f->parse(symName);
+    // This order, first adding to `bitcodeFiles` and then parsing is necessary.
+    // See https://github.com/llvm/llvm-project/pull/73095
     bitcodeFiles.push_back(f);
+    f->parse(symName);
```
Comment (sbc100) [inline lld/wasm/SymbolTable.cpp:56]: This looks like a good change to me.   The fact that ELF does it this way makes me even more confident this is correct: https://github.com/llvm/llvm-project/blob/3a6f02a6581b49b269710eea944dc114166403ed/lld/ELF/InputFiles.cpp#L314-L315

File: lld/test/wasm/lto/Inputs/global_var_init1.ll [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+target triple = "wasm32-unknown-unknown"
+
+; Generated from this C++ code:
+;
+; int foo();
+; inline int unused = foo();
+;
+; int main() {
+;   return foo();
+; }
```
Comment (sbc100) [inline lld/test/wasm/lto/Inputs/global_var_init1.ll:11]: I wonder if its possible to reduce these test files something smaller by hand writing them instead?

I'd always prefer something more concise in the test files in possible.

Also, I wonder if we even need to mention `global_var_init` .. presumably this problem applies to all comat groups?  Should we call these something `comdat_ordering.ll`?

**Review by sbc100 — COMMENTED** [review]
Thanks for debugging this issue!   It must have been a real pain to track down.   Nice to see such as simple fix though.

File: lld/test/wasm/lto/Inputs/global_var_init1.ll (hunk 2 above)
Comment (aheejin) [reply lld/test/wasm/lto/Inputs/global_var_init1.ll:11]: > I wonder if its possible to reduce these test files something smaller by hand writing them instead?
> 
> I'd always prefer something more concise in the test files in possible.

It was generated from that C++ code but hand-simplified later to delete some unnecessary parts. I think I can delete some more, but the overall structure is still the same. (That `@llvm.global_ctors` is still necessary; without this the error does not occur.) I think I can delete this comment.

> Also, I wonder if we even need to mention `global_var_init` .. presumably this problem applies to all comat groups? Should we call these something `comdat_ordering.ll`?

I don't think this problem is specific to `global_var_init`, but that all bug reports we got contained it kind of makes me think that the condition this bug is triggered is mostly created by `global_var_init`. I think I can rename the tests, but I guess it is fine to leave the variables in the tests as `__cxx_global_var_init`.

Comment (aheejin) [reply lld/test/wasm/lto/Inputs/global_var_init1.ll:11]: Hmm, no, turns out I can't really simplify the contents of `@__cxx_global_var_init` much. Even if I try to simplify it a little, the unused contents will be somehow deleted and `@__cxx_global_var_init` itself is deleted. Will rename the tests, and clarify the comment to say it is simplified after generation. (I think it is still better to have what it is generated from)

Comment (sbc100) [reply lld/test/wasm/lto/Inputs/global_var_init1.ll:11]: SGTM

**Review by sbc100 — APPROVED** [review]

**Comment by aheejin** [comment]
> (when landing maybe just include the short description?)

Sounds good

**Comment by nico** [comment]
Does doing something like #67445 instead help as well? If so, maybe go with that instead.

**Comment by sbc100** [comment]
> Does doing something like #67445 instead help as well? If so, maybe go with that instead.

This change seems like a clear improvement/bugfix to me.   If something like #67445 is needed then that would probably be for a separate issue, which I guess we have not yet seen?

**Comment by nico** [comment]
> This change seems like a clear improvement/bugfix to me. If something like #67445 is needed then that would probably be for a separate issue, which I guess we have not yet seen?

I think it prevents the "can lead to parsing of an object file, which in turn lead to parsing of another object file before finishing parsing of the previous file" issue mentioned further up, or at least limits it a bit. So maybe it helps here too. Not sure though.
